lib/AST/LocalizationFormat.cpp

```diff
@@ -143,30 +137,48 @@ void YAMLLocalizationProducer::forEachAvailable(
   }
 }
 
+llvm::Optional<uint32_t> getDiagnosticID(llvm::StringRef rawID) {
+  llvm::Optional<uint32_t> id;
+  auto it = llvm::find_if(diagnosticID,
+                          [&](const char *diagID) { return rawID == diagID; });
+  if (it != std::end(diagnosticID)) {
+    uint32_t index = static_cast<uint32_t>(std::distance(diagnosticID, it));
+    id.emplace(index);
+  }
+  return id;
+}
+
 template <typename T, typename Context>
 typename std::enable_if<llvm::yaml::has_SequenceTraits<T>::value, void>::type
-readYAML(llvm::yaml::IO &io, T &Seq, bool, Context &Ctx) {
+readYAML(llvm::yaml::IO &io, T &Seq, T &unknownIDs, bool, Context &Ctx) {
   unsigned count = io.beginSequence();
-  if (count)
+  if (count) {
     Seq.resize(LocalDiagID::NumDiags);
+  }
+
+  DiagnosticNode current;
   for (unsigned i = 0; i < count; ++i) {
     void *SaveInfo;
     if (io.preflightElement(i, SaveInfo)) {
-      DiagnosticNode current;
       yamlize(io, current, true, Ctx);
-      io.postflightElement(SaveInfo);
 
-      // A diagnostic ID might be present in YAML and not in `.def` file,
-      // if that's the case ScalarEnumerationTraits will assign the diagnostic ID
-      // to `LocalDiagID::NumDiags`. Since the diagnostic ID isn't available
-      // in `.def` it shouldn't be stored in the diagnostics array.
-      if (current.id != LocalDiagID::NumDiags) {
+      auto id = getDiagnosticID(current.id);
+      // A diagnostic ID might be present in YAML and not in `.def` file, if
+      // that's the case `id` will not have a value. Since the diagnostic ID
+      // isn't available in `.def` it shouldn't be stored in the diagnostics
+      // array.
+      if (id) {
+        io.postflightElement(SaveInfo);
+
         // YAML file isn't guaranteed to have diagnostics in order of their
         // declaration in `.def` files, to accommodate that we need to leave
         // holes in diagnostic array for diagnostics which haven't yet been
         // localized and for the ones that have `DiagnosticNode::id`
         // indicates their position.
-        Seq[static_cast<unsigned>(current.id)] = std::move(current.msg);
+        Seq[id.getValue()] = std::move(current.msg);
+      } else {
+        io.postflightElement(SaveInfo);
```

Comment:
It looks like `io.postflightElement(SaveInfo);` could be moved up from `if` condition...